src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -172,74 +171,63 @@ private static object[] ParseArguments(string methodIdentifier, string argsJson,
                 return Array.Empty<object>();
             }
 
-            // There's no direct way to say we want to deserialize as an array with heterogenous
-            // entry types (e.g., [string, int, bool]), so we need to deserialize in two phases.
-            var jsonDocument = JsonDocument.Parse(argsJson);
-            var shouldDisposeJsonDocument = true;
-            try
+            var utf8JsonBytes = Encoding.UTF8.GetBytes(argsJson).AsSpan();
+            var reader = new Utf8JsonReader(utf8JsonBytes);
+            if (!reader.Read() || reader.TokenType != JsonTokenType.StartArray)
             {
-                if (jsonDocument.RootElement.ValueKind != JsonValueKind.Array)
-                {
-                    throw new ArgumentException($"Expected a JSON array but got {jsonDocument.RootElement.ValueKind}.");
-                }
-
-                var suppliedArgsLength = jsonDocument.RootElement.GetArrayLength();
+                throw new JsonException("Invalid JSON");
+            }
 
-                if (suppliedArgsLength != parameterTypes.Length)
-                {
-                    throw new ArgumentException($"In call to '{methodIdentifier}', expected {parameterTypes.Length} parameters but received {suppliedArgsLength}.");
-                }
+            // Check if we have the right number of tokens
+            var suppliedArgsLength = 0;
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+            {
+                suppliedArgsLength++;
+                reader.Skip();
```

Comment:
nit: I think this can be written a bit differently to avoid the double parsing of the json data.

Currently we are doing one pass to validate, and then the other to do the actual work.

Reply:
I agree, this does seem a bit odd

Reply:
Also maybe I'm misunderstanding the APIs, but does `reader.Skip` skip over complex objects or does it just skip over individual tokens? How many tokens does it think are in the input `["a", ["b1", "b2"]]`, or is that sort of thing not possible here?

Reply:
`reader.Skip` skips over complex objects (not individual tokens).

So, if your reader was positioned at StartArray (or Object), it would skip to EndArray (or Object). If your reader was positioned at String, or other primitive token (like Null, True, False, Number), it wouldn't do anything.